lib/ex_plasma/transaction/type/payment_v1.ex

```diff
@@ -1,150 +1,251 @@
+  def to_rlp(%PaymentV1{} = transaction) do
+    %PaymentV1{inputs: inputs, outputs: outputs, metadata: metadata} = transaction
+
     [
-      transaction.sigs,
       <<@tx_type>>,
-      Enum.map(transaction.inputs, &ExPlasma.Output.to_rlp_id/1),
-      Enum.map(transaction.outputs, &ExPlasma.Output.to_rlp/1),
+      Enum.map(inputs, &Output.to_rlp_id/1),
+      Enum.map(outputs, &Output.to_rlp/1),
       @empty_tx_data,
-      transaction.metadata || @empty_metadata
+      metadata || @empty_metadata
     ]
   end
 
   @doc """
   Decodes an RLP list into a Payment V1 Transaction.
 
-  ## Example
-
-  iex> rlp = [
-  ...>  [], 
-  ...>  <<1>>, 
-  ...>  [<<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>], 
-  ...>  [
-  ...>    [
-  ...>      <<1>>, 
-  ...>      [<<29, 246, 47, 41, 27, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>, <<46, 38, 45, 41, 28, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>, <<1>>]
-  ...>    ]
-  ...>  ], 
-  ...>  0,
-  ...>  <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>
-  ...>]
-  iex> ExPlasma.Transaction.Type.PaymentV1.to_map(rlp)
-  %ExPlasma.Transaction{
-  	inputs: [
-  		%ExPlasma.Output{
-  			output_data: nil,
-  			output_id: %{blknum: 0, oindex: 0, position: 0, txindex: 0},
-  			output_type: nil
-  		}
-  	],
-  	metadata: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
-  	outputs: [
-  		%ExPlasma.Output{
-  			output_data: %{
-  				amount: 1,
-  				output_guard: <<29, 246, 47, 41, 27, 46, 150, 159, 176, 132, 157, 153,
-  					217, 206, 65, 226, 241, 55, 0, 110>>,
-  				token: <<46, 38, 45, 41, 28, 46, 150, 159, 176, 132, 157, 153, 217, 206,
-  					65, 226, 241, 55, 0, 110>>
-  			},
-  			output_id: nil,
-  			output_type: 1
-  		}
-  	],
-  	sigs: [],
-  	tx_data: 0,
-  	tx_type: 1
-  }
+  Only validates that the RLP is structurally correct.
+  Does not perform any other kind of validation, use validate/1 for that.
   """
-  @impl Transaction
-  @spec to_map(list()) :: Transaction.t()
-  def to_map(rlp) when is_list(rlp), do: do_to_map(rlp)
-
-  defp do_to_map([sigs, tx_type, inputs, outputs, "", metadata]),
-    do: do_to_map([sigs, tx_type, inputs, outputs, 0, metadata])
-
-  defp do_to_map([sigs, <<tx_type>>, inputs, outputs, tx_data, metadata]),
-    do: do_to_map([sigs, tx_type, inputs, outputs, tx_data, metadata])
-
-  defp do_to_map([sigs, tx_type, inputs, outputs, tx_data, metadata]) do
-    %ExPlasma.Transaction{
-      sigs: sigs,
-      tx_type: tx_type,
-      inputs: Enum.map(inputs, &ExPlasma.Output.decode_id/1),
-      outputs: Enum.map(outputs, &ExPlasma.Output.decode/1),
-      tx_data: tx_data,
-      metadata: metadata
-    }
+  @spec to_map(PaymentV1.t(), list()) :: {:ok, PaymentV1.t()} | {:error, mapping_error()}
+  def to_map(%PaymentV1{}, [<<@tx_type>>, inputs_rlp, outputs_rlp, tx_data_rlp, metadata_rlp]) do
```

Comment:
since you're in a protocol already, you do not need to type pattern match in the function header `%PaymentV1{}`